src/tdyti.c

```diff
@@ -104,51 +104,65 @@ PetscErrorCode TDyTimeIntegratorRunToTime(TDy tdy,PetscReal sync_time) {
+            break;
+          case TH:
+            ierr = PetscPrintf(PETSC_COMM_WORLD,"===== TH MODE ====================================\n"); CHKERRQ(ierr);
+            break;
+        }
         ierr = TDyTimeIntegratorSetTargetTime(ti,sync_time); CHKERRQ(ierr);
         ierr = TDySetDtimeForSNESSolver(tdy,ti->dt); CHKERRQ(ierr);
-        if (!rank){
-          switch (tdy->options.mode){
-            case RICHARDS:
-              printf("===== RICHARDS MODE ==============================\n");
-              break;
-            case TH:
-              printf("===== TH MODE ====================================\n");
-              break;
+        solver_converged = PETSC_FALSE;
+        while (!solver_converged){
+          ierr = TDyPreSolveSNESSolver(tdy); CHKERRQ(ierr);
+          ierr = SNESSolve(ti->snes,PETSC_NULL,tdy->soln); CHKERRQ(ierr);
+          PetscInt nit, lit;
+          ierr = SNESGetIterationNumber(ti->snes,&nit); CHKERRQ(ierr);
+          ierr = SNESGetLinearSolveIterations(ti->snes,&lit); CHKERRQ(ierr);
+          ierr = SNESGetConvergedReason(ti->snes,&reason); CHKERRQ(ierr);
+          if (reason < 0) {
+            icut++;
+            ti->dt *= 0.25;
+            ierr = TDySetDtimeForSNESSolver(tdy,ti->dt); CHKERRQ(ierr);
+            PetscChar message[1000];
+            sprintf(message,"-> Cut time step: snes=%d icut=%d time=%12.5e dt=%12.5e\n",reason,icut,ti->time,ti->dt);
+            ierr = PetscPrintf(PETSC_COMM_WORLD,message);
+            ierr = TDyTimeCut(tdy); CHKERRQ(ierr);
+            if (icut == max_time_step_cuts) {
+              SETERRQ(PETSC_COMM_WORLD,PETSC_ERR_USER,"TDyTimeIntegratorRunToTime failed to converge after maximum time step cuts");
+            }
+          } else {
+            solver_converged = PETSC_TRUE;
+            ti->time += ti->dt;
+            ti->istep++;
+            ierr = TDyPostSolve(tdy,tdy->soln); CHKERRQ(ierr);
           }
-        }
-        ierr = TDyPreSolveSNESSolver(tdy); CHKERRQ(ierr);
-        ierr = SNESSolve(ti->snes,PETSC_NULL,tdy->soln); CHKERRQ(ierr);
-        ierr = TDyPostSolve(tdy,tdy->soln); CHKERRQ(ierr);
-        ti->time += ti->dt;
-        ti->istep++;
-        PetscInt nit, lit;
-        ierr = SNESGetIterationNumber(ti->snes,&nit); CHKERRQ(ierr);
-        ierr = SNESGetLinearSolveIterations(ti->snes,&lit); CHKERRQ(ierr);
-        ierr = SNESGetConvergedReason(ti->snes,&reason); CHKERRQ(ierr);
-        if (reason < 0) {
-          SETERRQ(PETSC_COMM_WORLD,PETSC_ERR_USER,"TDyTimeIntegratorRunToTime failed to converge");
-        }
 
-        if (tdy->io->io_process) {
-          printf("\nTime step %d: time = %f dt = %f ni = %d li = %d rsn = %s\n\n",
+          PetscChar message[1000];
+          sprintf(message,"\nTime step %d: time = %f dt = %f ni = %d li = %d rsn = %s\n\n",
             ti->istep,ti->time,ti->dt,nit,lit, SNESConvergedReasons[reason]);
+          ierr = PetscPrintf(PETSC_COMM_WORLD,message); CHKERRQ(ierr);
```

Comment:
Actually, [PetscPrintf](https://petsc.org/release/docs/manualpages/Sys/PetscPrintf.html) allows you to specify formatting parameters, just like `printf` does. Why not just replace your previous call to `printf` with `PetscPrintf`?